accounts/scwallet/wallet.go

```diff
@@ -0,0 +1,1071 @@
+			}
+			// Pairing still ok, fall through to PIN checks
+		} else {
+			// If no passphrase was supplied, request the PUK from the user
+			if passphrase == "" {
+				return ErrPairingPasswordNeeded
+			}
+			// Attempt to pair the smart card with the user supplied PUK
+			if err := w.pair([]byte(passphrase)); err != nil {
+				return err
+			}
+			// Pairing succeeded, fall through to PIN checks. This will of course fail,
+			// but we can't return ErrPINNeeded directly here becase we don't know whether
+			// a PIN check or a PIN reset is needed.
+			passphrase = ""
+		}
+	}
+	// The smart card was successfully paired, retrieve its status to check whether
+	// PIN verification or unblocking is needed.
+	status, err := w.session.walletStatus()
+	if err != nil {
+		return err
+	}
+	// Request the appropriate next authentication data, or use the one supplied
+	switch {
+	case passphrase == "" && status.PinRetryCount > 0:
+		return ErrPINNeeded
+	case passphrase == "":
+		return ErrPINUnblockNeeded
+	case status.PinRetryCount > 0:
+		if err := w.session.verifyPin([]byte(passphrase)); err != nil {
+			return err
+		}
+	default:
+		if err := w.session.unblockPin([]byte(passphrase)); err != nil {
+			return err
+		}
+	}
+	// Smart card paired and unlocked, initialize and register
+	w.deriveReq = make(chan chan struct{})
+	w.deriveQuit = make(chan chan error)
+
+	go w.selfDerive()
+
+	// Notify anyone listening for wallet events that a new device is accessible
+	go w.Hub.updateFeed.Send(accounts.WalletEvent{Wallet: w, Kind: accounts.WalletOpened})
+
+	return nil
+}
+
+// Close stops and closes the wallet, freeing any resources.
+func (w *Wallet) Close() error {
+	// Ensure the wallet was opened
+	w.lock.Lock()
+	dQuit := w.deriveQuit
+	w.lock.Unlock()
+
+	// Terminate the self-derivations
+	var derr error
+	if dQuit != nil {
+		errc := make(chan error)
+		dQuit <- errc
+		derr = <-errc // Save for later, we *must* close the USB
+	}
+	// Terminate the device connection
+	w.lock.Lock()
+	defer w.lock.Unlock()
+
+	w.deriveQuit = nil
+	w.deriveReq = nil
+
+	if err := w.release(); err != nil {
+		return err
+	}
+	return derr
+}
+
+// selfDerive is an account derivation loop that upon request attempts to find
+// new non-zero accounts.
+func (w *Wallet) selfDerive() {
+	w.log.Debug("Smart card wallet self-derivation started")
+	defer w.log.Debug("Smart card wallet self-derivation stopped")
+
+	// Execute self-derivations until termination or error
+	var (
+		reqc chan struct{}
+		errc chan error
+		err  error
+	)
+	for errc == nil && err == nil {
+		// Wait until either derivation or termination is requested
+		select {
+		case errc = <-w.deriveQuit:
+			// Termination requested
+			continue
+		case reqc = <-w.deriveReq:
+			// Account discovery requested
+		}
+		// Derivation needs a chain and device access, skip if either unavailable
+		w.lock.Lock()
+		if w.session == nil || w.deriveChain == nil {
```

Comment:
Does it really require chain access? For clef, it won't have that. Clef instead tells the hw wallets to derive 5 (or 10, can't remember) accounts. 
Would be good if it could use this method, perhaps if it takes a `numToDerive` as a parameter, which can be set to `0` for chain-backed self derivation?